devnet/docker-compose.yml

```diff
@@ -0,0 +1,65 @@
+services: 
+  alice:
+    build:
+      context: ./..
+      dockerfile: devnet/alice.Dockerfile
+      args:
+        DD_API_KEY: ${DD_API_KEY}
+        pubip: 10.5.0.2
```

Comment:
Can't the program auto-detect it's LAN IP address?

Reply:
We use a library called `pubip` which requests from the internet from a bunch of those "what's my ip" services.  The `pubip` key I introduced forces the node to broadcast from the specified address.  I had some code that scanned all network interfaces, but if you have multiple addresses which one would you pick?

Reply:
For LAN ip addresses, you can `go get github.com/vishvananda/netlink` and use

```go
func Test_findLANIP(t *testing.T) {
	ip, err := findLANIP()
	require.NoError(t, err)
	t.Log(ip)
}

var (
	ErrNoLinkFound            = errors.New("no link was found")
	ErrDefaultGatewayNotFound = errors.New("default gateway not found")
)

func findLANIP() (ip net.IP, err error) {
	links, err := netlink.LinkList()
	if err != nil {
		return nil, fmt.Errorf("cannot list links: %w", err)
	} else if len(links) == 0 {
		return nil, ErrNoLinkFound
	}

	for _, link := range links {
		routes, err := netlink.RouteList(link, netlink.FAMILY_ALL)
		if err != nil {
			return nil, fmt.Errorf(
				"cannot list routes for link %d: %w",
				link.Attrs().Index, err)
		}

		for _, route := range routes {
			if route.Dst == nil {
				return route.Gw, nil
			}
		}
	}

	return nil, ErrDefaultGatewayNotFound
}
```

Although that's limited to unix systems and introduces an indirect dependency on syscall (so for windows we would need different build tags etc.)